Library/Homebrew/cli/named_args.rb

```diff
@@ -29,13 +42,212 @@ def initialize(*args, parent: Args.new, override_spec: nil, force_bottle: false,
         super(@args)
       end
 
+      # The expected type of the named arguments.
+      attr_accessor :type
+
       attr_reader :parent
 
+      # Convert named arguments to the type defined by {.type}.
+      # This method returns a list of converted named arguments.
+      # If `separate_types` is true, the method will return a hash containing
+      # the keys `:formula`, `:cask`, `:keg`, `:other`. Each key points to an array
+      # containing the converted named args of that type. An exception will be raised
+      # if the args are unable to be converted to the appropriate type
+      # unless `ignore_unavailable` is true.
+      # @return [Array] if `separate_types` is `false`
+      # @return [Hash] if `separate_types` is `true`
+      # @api public
+      sig do
+        params(
+          only:                         T.nilable(Symbol),
+          separate_types:               T::Boolean,
+          ignore_unavailable:           T::Boolean,
+          allow_formula_cask_conflicts: T::Boolean,
+        ).returns(T.any(Array, Hash))
+      end
+      def convert(only: nil, separate_types: false, ignore_unavailable: false, allow_formula_cask_conflicts: false)
```

Comment:
`to_objects` or something feels like a nicer method name. I wonder if `to_formulae_and_casks` or something could be used and just change the API?

Also, this method feels really big; maybe it's worth breaking down a bit?

Reply:
The reason I wanted to move away from `to_*` was that I wanted this method to be a catch-all. One could _always_ call `args.named.convert` and get the converted objects. Maybe that's not the way to go about this.

I'd be in favor of breaking it down a little, but I think the main reason it's so long is it has a huge `case` statement. It already calls out to helper methods for the actual arg conversion.

Reply:
I feel like having the return type of this depend on what is passed to `CLI::Parser` is too implicit (also no way to have a meaningful type signature).

Maybe we could store argument as `Arg` objects which have e.g. a `to_cask` method, etc., so the API would look like:

```ruby
args.named.map(&:to_cask)
```

Reply:
> I feel like having the return type of this depend on what is passed to `CLI::Parser` is too implicit (also no way to have a meaningful type signature).

Agreed 👍🏻 

> Maybe we could store arguments as `Arg` objects which have e.g. a `to_cask` method, etc., so the API would look like:
> 
> ```ruby
> args.named.map(&:to_cask)
> ```

Yeh, something like this looks nice. I'm basically wanting to have "if I want to get casks the call method/parameters contain the word `cask`"

Reply:
Agreed, something like this is a better solution. Let me give some more thought to this and whether this is actually any better than the existing `args.named.to_*` methods. We were using those successfully before so I don't think we need to reinvent them just to allow this `named_args` method.

_That's why this is a draft PR. I'm still working through everything and figuring out what the best way to do it is_

Reply:
> _That's why this is a draft PR. I'm still working through everything and figuring out what the best way to do it is_

Yup, all good ❤️

Reply:
> We were using those successfully before so I don't think we need to reinvent them just to allow this `named_args` method.

Potentially they could be extended to `raise` if called in the wrong way (i.e. if `named_args :tap` and you request `named.to_formulae` or something?).